beacon-chain/sync/initial-sync/round_robin.go

```diff
@@ -0,0 +1,259 @@
+package initialsync
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/paulbellamy/ratecounter"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	prysmsync "github.com/prysmaticlabs/prysm/beacon-chain/sync"
+	"github.com/prysmaticlabs/prysm/beacon-chain/sync/peerstatus"
+	p2ppb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	eth "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+)
+
+const blockBatchSize = 64
+const counterSeconds = 20
+
+// Round Robin sync looks at the latest peer statuses and syncs with the highest
+// finalized peer,
+//
+// Step 1 - Sync to finalized epoch.
+// Sync with peers of lowest finalized root with epoch greater than head state.
+//
+// Step 2 - Sync to head from finalized epoch.
+// Using the finalized root as the head_block_root and the epoch start slot
+// after the finalized epoch, request blocks to head from some subset of peers
+// where step = 1.
+func (s *InitialSync) roundRobinSync(genesis time.Time) error {
+	ctx := context.Background()
+
+	counter := ratecounter.NewRateCounter(counterSeconds * time.Second)
+
+	var requestBlocks = func(req *p2ppb.BeaconBlocksByRangeRequest, pid peer.ID) ([]*eth.BeaconBlock, error) {
+		log.WithField("peer", pid.Pretty()).WithField("req", req).Debug("requesting blocks")
+		stream, err := s.p2p.Send(ctx, req, pid)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to send request to peer")
+		}
+		defer stream.Close()
+
+		resp := make([]*eth.BeaconBlock, 0, req.Count)
+		for {
+			blk, err := prysmsync.ReadChunkedBlock(stream, s.p2p)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to read chunked block")
+			}
+			resp = append(resp, blk)
+		}
+
+		return resp, nil
+	}
+
+	// Step 1 - Sync to end of finalized epoch.
+	for s.chain.HeadSlot() < helpers.StartSlot(highestFinalizedEpoch()+1) {
+		root, finalizedEpoch, peers := highestFinalized()
+
+		var blocks []*eth.BeaconBlock
+
+		// request a range of blocks to be requested from multiple peers.
+		// Example:
+		//   - number of peers = 4
+		//   - range of block slots is 64...128
+		//   Four requests will be spread across the peers using step argument to distribute the load
+		//   i.e. the first peer is asked for block 64, 68, 72... while the second peer is asked for
+		//   65, 69, 73... and so on for other peers.
+		var request func(start uint64, step uint64, count uint64, peers []peer.ID, remainder int) ([]*eth.BeaconBlock, error)
+		request = func(start uint64, step uint64, count uint64, peers []peer.ID, remainder int) ([]*eth.BeaconBlock, error) {
```

Comment:
Also this can algorthim can maybe be separated into its own method

Reply:
The thing about moving this outside of the scope of this method is that it would require many arguments. It uses the finalized root, blocks array pointer, and finalizedEpoch reference. 

I think its OK to leave how we have it

Reply:
ah k , yeah then its best to leave it as it is